EOS-15971: Modify s3_setup shell script to Python - Cleanup util

pratyush-seagate opened the pull request:
commit 38ea9773040a74dd28855b0e19189f66beb54f41
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Fri Jan 8 09:33:20 2021 -0700

            EOS-15971: Modify s3_setup shell script to Python - Cleanup util
            Change description:
              1. Integrate the newly added s3psetup with build system so that s3server's rpm will contain this program.
              2. Integrate s3psetup with s3_setup only for cleanup operation. Note to reviewers: Dependency cleanup causes config files to be removed. This causes additional problems to the expected idempotency nature of the control flow in s3-setup. Hence dependency cleanup is temporarily disabled in s3_setup.
              3. Update s3psetup, bug fixes found during unit tests, log addition, use optional ldappasswd etc.
              4. Add sample test scripts used to unit test this change. These scripts also contain additional steps so that s3_cleanup can run again and again w/o any failures. (ref. clean_open_ldap_by_s3.sh, prerequisite to test_s3_setup.sh)
            List of files:
                modified:   rpms/s3/s3rpm.spec
                modified:   s3psetup/README.md
                modified:   s3psetup/s3psetup/cortx_s3_py_setup.py
                modified:   scripts/provisioning/s3_setup
                new file:   clean_open_ldap_by_s3.sh
                new file:   test_s3_setup.sh
            Testing:
            UT:
              On single node VM, based on the steps provided by Amit, created test_s3_setup.sh and ran it multiple times with this rpm. Added debug logs were also checked to confirm there were no failures.

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

commit 572bafd4217c22d6a2d7b3aa72ffb7c799959ecb
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Wed Jan 6 04:26:01 2021 -0700

        EOS-15971: Modify s3_setup shell script to Python - Cleanup util
        Change description:
        Introduce new python version of the existing s3_setup script and add the cleanup utility.
        List of files:
                new file:   README.md
                new file:   VERSION
                new file:   s3psetup/__init__.py
                new file:   s3psetup/cortx_s3_py_setup.py
                new file:   s3psetup/s3psetup
                new file:   setup.py
        Testing:
        UT:
                1. For unit test, on a single VM:
                a. run ./scripts/env/dev/./init.sh from s3 repo (ref. https://github.com/Seagate/cortx-s3server/blob/main/docs/CORTX-S3%20Server%20Quick%20Start%20Guide.md#12-installing-dependencies)
                b. ./jenkins-build.sh --skip_ut_build --skip_tests --skip_ut_run --skip_st_run to do a setup. Additional options are passed to reduce the setup time as that script runs many tests by default.
                c. run "/opt/seagate/cortx/s3/bin/s3_setup --cleanup" with control passed to the newly created s3psetup binary for the cleanup and compare the step results to runs without using the new binary. Note: s3_setup's variable "ldappasswd" needs to set explicitly at line number 310 otherwise the password does not match and script fails.
                There are few issues noticed during ut. ldapsearch confirms records are found, after this access key is being fetch from s3cipher but the follow up ldapdelete using that accesskey is failing with error: ldap_delete: No such object (32)

                2. For integration test with provisioner:
                For testing "s3_setup --cleanup" changes in s3_setup script which is used by provisioner, I was trying to create a single node cluster using provisioner dev environment setup, and was planning to replace the /opt/seagate/cortx/s3/bin/s3_setup with my custom s3_setup (since provisioner is not allowing me to install my custom rpms or I'm not aware of how to do this, so I'm following this script replace approach). Now I need to run provisioner cleanup for s3 or whole cluster, but not sure what will be the provisioner command to trigger this. Requested help from teammates on this point and waiting for their input.

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

pratyush-seagate commented:
Current debug enabled output from the new s3psetup binary:

[root@ssc-vm-1733 s3psetup]# s3psetup --cleanup both --ldappasswd ldapadmin
1: ldapsearch -b "o=s3-background-delete-svc,ou=accounts,dc=s3,dc=seagate,dc=com" -x -w ldapadmin -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldap://
2: # extended LDIF
#
# LDAPv3
# base <o=s3-background-delete-svc,ou=accounts,dc=s3,dc=seagate,dc=com> with scope subtree
# filter: (objectclass=*)
# requesting: ALL
#

# search result
search: 2
result: 32 No such object
matchedDN: ou=accounts,dc=s3,dc=seagate,dc=com

# numResponses: 1

3: ['#', 'extended', 'LDIF\n#\n#', 'LDAPv3\n#', 'base', '<o=s3-background-delete-svc,ou=accounts,dc=s3,dc=seagate,dc=com>', 'with', 'scope', 'subtree\n#', 'filter:', '(objectclass=*)\n#', 'requesting:', 'ALL\n#\n\n#', 'search', 'result\nsearch:', '2\nresult:', '32', 'No', 'such', 'object\nmatchedDN:', 'ou=accounts,dc=s3,dc=seagate,dc=com\n\n#', 'numResponses:', '1\n']
7: ldapdelete -x -w ldapadmin -r "ak=NF9ZWTk3Nk5RN3ZjaDBaZW,ou=accesskeys,dc=s3,dc=seagate,dc=com" -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldapi:///
8:  ldap_search: No such object (32)
ldap_delete: No such object (32)
        matched DN: ou=accesskeys,dc=s3,dc=seagate,dc=com

9: ldapdelete -x -w ldapadmin -r "o=s3-background-delete-svc,ou=accounts,dc=s3,dc=seagate,dc=com" -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldapi:///
10:  ldap_search: No such object (32)
ldap_delete: No such object (32)
        matched DN: ou=accounts,dc=s3,dc=seagate,dc=com

11: Deleted s3backgrounddelete account successfully...
12: ldapsearch -b "o=s3-recovery-svc,ou=accounts,dc=s3,dc=seagate,dc=com" -x -w ldapadmin -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldap://
13: # extended LDIF
#
# LDAPv3
# base <o=s3-recovery-svc,ou=accounts,dc=s3,dc=seagate,dc=com> with scope subtree
# filter: (objectclass=*)
# requesting: ALL
#

# search result
search: 2
result: 32 No such object
matchedDN: ou=accounts,dc=s3,dc=seagate,dc=com

# numResponses: 1

17: ldapdelete -x -w ldapadmin -r "ak=d3I4UGpjbmpzbUNyN1dMam,ou=accesskeys,dc=s3,dc=seagate,dc=com" -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldapi:///
18:  ldap_search: No such object (32)
ldap_delete: No such object (32)
        matched DN: ou=accesskeys,dc=s3,dc=seagate,dc=com

19: ldapdelete -x -w ldapadmin -r "o=s3-recovery-svc,ou=accounts,dc=s3,dc=seagate,dc=com" -D "cn=sgiamadmin,dc=seagate,dc=com" -H ldapi:///
20:  ldap_search: No such object (32)
ldap_delete: No such object (32)
        matched DN: ou=accounts,dc=s3,dc=seagate,dc=com

21: Deleted s3recovery tool account successfully...
removing s3 dependencies
Loaded plugins: fastestmirror, priorities, product-id, search-disabled-repos,
              : subscription-manager
No Packages marked for removal
 Repository releases_cortx_s3deps is listed more than once in the configuration
No Match for argument: log4cxx_cortx

 mv: missing destination file operand after ‘/opt/seagate/cortx/auth/resources/authserver.properties’
Try 'mv --help' for more information.

 mv: cannot stat ‘/opt/seagate/cortx/s3/s3backgrounddelete/config.yaml’: No such file or directory

 mv: cannot stat ‘/opt/seagate/cortx/auth/resources/keystore.properties’: No such file or directory

 mv: cannot stat ‘/opt/seagate/cortx/s3/conf/s3config.yaml’: No such file or directory

pratyush-seagate commented:
[root@ssc-vm-1735 ~]#  ./test_setup.sh
slapd could not be stopped!
s3:post_install passed!
s3:config failed!
[root@ssc-vm-1735 ~]#  ./test_setup.sh
[root@ssc-vm-1735 ~]#  ./test_setup.sh
Loaded plugins: fastestmirror, product-id, search-disabled-repos, subscription-manager
Resolving Dependencies
--> Running transaction check
---> Package openldap-clients.x86_64 0:2.4.44-22.el7 will be erased
---> Package openldap-servers.x86_64 0:2.4.44-22.el7 will be erased
--> Finished Dependency Resolution

Dependencies Resolved

======================================================================================================================================================
 Package                              Arch                       Version                           Repository                                     Size
=======================================================================================================================================================
Removing:
 openldap-clients                     x86_64                     2.4.44-22.el7                     @EOS_CentOS-7_CentOS-7-OS                     570 k
 openldap-servers                     x86_64                     2.4.44-22.el7                     @EOS_CentOS-7_CentOS-7-OS                     5.1 M

Transaction Summary
=======================================================================================================================================================
Remove  2 Packages

Installed size: 5.6 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Erasing    : openldap-servers-2.4.44-22.el7.x86_64                                                                                               1/2
  Erasing    : openldap-clients-2.4.44-22.el7.x86_64                                                                                               2/2
Loaded plugins: fastestmirror, product-id, subscription-manager
  Verifying  : openldap-clients-2.4.44-22.el7.x86_64                                                                                               1/2
  Verifying  : openldap-servers-2.4.44-22.el7.x86_64                                                                                               2/2
EOS_CentOS-7_CentOS-7-Extras                                                                                                    | 2.1 kB  00:00:00
EOS_CentOS-7_CentOS-7-OS                                                                                                        | 2.1 kB  00:00:00
EOS_CentOS-7_CentOS-7-Updates                                                                                                   | 2.1 kB  00:00:00
EOS_EPEL-7_EPEL-7                                                                                                               | 2.5 kB  00:00:00
EOS_Puppet6_puppet6-el7                                                                                                         | 2.1 kB  00:00:00

Removed:
  openldap-clients.x86_64 0:2.4.44-22.el7                                    openldap-servers.x86_64 0:2.4.44-22.el7

Complete!
s3:post_install passed!
s3:config passed!
s3:init passed w/o --setlogrotate!
s3:cleanup passed!

pratyush-seagate commented:
Custom build in progress with this change at: http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/380/

jenkins job will be run after this build completes.

pratyush-seagate commented:
> 
> 
> Custom build in progress with this change at: http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/380/
> 
> jenkins job will be run after this build completes.

Update:
Custom build at http://cortx-storage.colo.seagate.com/releases/cortx/github/integration-custom-ci/release/centos-7.8.2003/custom-build-380/ tested on VM and rpm integration found ok.

Now Jenkins started at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/

amitwac2608 commented on s3psetup/README.md:
```diff
@@ -0,0 +1,28 @@
+For any questions about this software or licensing,
+please email opensource@seagate.com or cortx-questions@seagate.com.
+
+## How to build and install s3psetup:
```
what is s3psteup ? give one liner description. I would suggest to have script name as "s3setup"

Also I think to create the directory with name "s3setup" at scripts/provisioning/

And put all this python, readme files inside it.

pratyush-seagate replied:
Reason using name "s3psetup" to distinguish from current shell script "s3_setup" ("p for python version"). Restructure the scripts might take some time due to rerun the builds and jenkins jobs. Will that be ok?

amitwac2608 replied:
yes, change the name now to "s3setup"

amitwac2608 commented on s3psetup/clean_open_ldap_by_s3.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
```
copyright header.

I guess this script is going to be temporary.. Please add this in comment.

We can have name of the file as : s3openldap_config_cleanup.py !

pratyush-seagate replied:
Considering this is a temporary script, can we let it be a shell script for the time being? Converting it to python is additional work and I feel we may not need to spend time on this conversion as the base script is already written, working and tested. And this script is also temporary.

amitwac2608 replied:
keep it shell script, but add comment that this is temp file.

amitwac2608 commented on s3psetup/clean_open_ldap_by_s3.sh:
```diff
@@ -0,0 +1,16 @@
+
+# Attempt ldap clean up since ansible openldap setup is not idempotent
+systemctl stop slapd 2>/dev/null || die_with_error "slapd could not be stopped!"
+yum remove -y openldap-servers openldap-clients || "openldap-servers openldap-clients removal failed"
```
we decided to not un-install the openldap packages, right ? @AjinkyaDhumal !

pratyush-seagate replied:
After those slapd config files are removed, slapd fails to start.

[root@ssc-vm-1735 ~]# systemctl stop slapd
[root@ssc-vm-1735 ~]# systemctl start slapd
Job for slapd.service failed because a configured resource limit was exceeded. See "systemctl status slapd.service" and "journalctl -xe" for details.
[root@ssc-vm-1735 ~]# systemctl status slapd.service
● slapd.service - OpenLDAP Server Daemon
   Loaded: loaded (/usr/lib/systemd/system/slapd.service; enabled; vendor preset: disabled)
   Active: failed (Result: resources) since Mon 2021-01-11 04:24:22 MST; 41s ago
     Docs: man:slapd
           man:slapd-config
           man:slapd-hdb
           man:slapd-mdb
           file:///usr/share/doc/openldap-servers/guide.html
  Process: 34221 ExecStart=/usr/sbin/slapd -u ldap -h ${SLAPD_URLS} $SLAPD_OPTIONS (code=exited, status=0/SUCCESS)
  Process: 34192 ExecStartPre=/usr/libexec/openldap/check-config.sh (code=exited, status=0/SUCCESS)
 Main PID: 34223 (code=exited, status=0/SUCCESS)

Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Stopped OpenLDAP Server Daemon.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to load environment files: No such file or directory
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed to run 'start-pre' task: No such file or directory
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to start OpenLDAP Server Daemon.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Unit slapd.service entered failed state.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed.
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to load environment files: No such file or directory
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed to run 'start-pre' task: No such file or directory
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to start OpenLDAP Server Daemon.
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed.


Those open ldap packages needs to be removed and reinstalled to get around this problem.

amitwac2608 commented on s3psetup/s3psetup/cortx_s3_py_setup.py:
```diff
@@ -0,0 +1,154 @@
+#
```
rename file to: cortx_s3_setup.py

AjinkyaDhumal replied:
I think you should add the s3psetup directory to cortx-s3server/scripts/provisioning

AjinkyaDhumal commented on s3psetup/clean_open_ldap_by_s3.sh:
```diff
@@ -0,0 +1,16 @@
+
+# Attempt ldap clean up since ansible openldap setup is not idempotent
+systemctl stop slapd 2>/dev/null || die_with_error "slapd could not be stopped!"
+yum remove -y openldap-servers openldap-clients || "openldap-servers openldap-clients removal failed"
```
Can you remove this line. Not required to remove the package during cleanup.

pratyush-seagate replied:
This script is not part of the production code. Last friday you and Amit suggested to keep the steps that needs to be run to cleanup ldap accounts in a separate script so that anyone needs to run cleanup and rerun setup can manually run the steps in between until s3_setup becomes completely idempotent.

AjinkyaDhumal replied:
yes agreed, but removing ldap packages is not required for cleanup purpose. This was tested I think. Just confirm once if you can re-configure ldap without removing the packages. If so we can remove that line.

pratyush-seagate replied:
After those slapd config files are removed, slapd fails to start.

[root@ssc-vm-1735 ~]# systemctl stop slapd
[root@ssc-vm-1735 ~]# systemctl start slapd
Job for slapd.service failed because a configured resource limit was exceeded. See "systemctl status slapd.service" and "journalctl -xe" for details.
[root@ssc-vm-1735 ~]# systemctl status slapd.service
● slapd.service - OpenLDAP Server Daemon
   Loaded: loaded (/usr/lib/systemd/system/slapd.service; enabled; vendor preset: disabled)
   Active: failed (Result: resources) since Mon 2021-01-11 04:24:22 MST; 41s ago
     Docs: man:slapd
           man:slapd-config
           man:slapd-hdb
           man:slapd-mdb
           file:///usr/share/doc/openldap-servers/guide.html
  Process: 34221 ExecStart=/usr/sbin/slapd -u ldap -h ${SLAPD_URLS} $SLAPD_OPTIONS (code=exited, status=0/SUCCESS)
  Process: 34192 ExecStartPre=/usr/libexec/openldap/check-config.sh (code=exited, status=0/SUCCESS)
 Main PID: 34223 (code=exited, status=0/SUCCESS)

Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Stopped OpenLDAP Server Daemon.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to load environment files: No such file or directory
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed to run 'start-pre' task: No such file or directory
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to start OpenLDAP Server Daemon.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: Unit slapd.service entered failed state.
Jan 11 04:24:22 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed.
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to load environment files: No such file or directory
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed to run 'start-pre' task: No such file or directory
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: Failed to start OpenLDAP Server Daemon.
Jan 11 04:24:58 ssc-vm-1735.colo.seagate.com systemd[1]: slapd.service failed.


Those open ldap packages needs to be removed and reinstalled to get around this problem.

sachitanands replied:
openldap-servers, openldap-clients rpms installs the slpad.service file and its service releated files ..if we are planning to start this slapd.service then these packages needs to be installed.

amitwac2608 commented on scripts/provisioning/s3_setup:
```diff
@@ -83,6 +83,7 @@ update_cluster_id=false
 is_rabbitmq_clustered=false
 cleanup_accounts=false
 cleanup_dependencies=false
+cleanup_both=false
```
cleanup_both is not good variable name.. use something which is more informative, like: "cleanup_ldap_accounts_and_dependencies"

pratyush-seagate replied:
Ajinkya suggested to use "cleanup" only. I'll go with that as it is short.

amitwac2608 commented on scripts/provisioning/s3_setup:
```diff
@@ -146,8 +154,9 @@ do
     --cleanup )
-        cleanup_accounts=true
-        cleanup_dependencies=true
+        # cleanup_accounts=true
```
remove commented code

AjinkyaDhumal commented on scripts/provisioning/s3_setup:
```diff
@@ -146,8 +154,9 @@ do
-        cleanup_dependencies=true
+        # cleanup_accounts=true
+        # cleanup_dependencies=true
+        cleanup_both=true
```
you can have just 'cleanup' instead of 'cleanup_both'

AjinkyaDhumal replied:
Better to have it as - 'cleanup_s3_accounts_schema'

amitwac2608 commented on s3psetup/s3psetup/cortx_s3_py_setup.py:
```diff
@@ -0,0 +1,154 @@
+
+  def run(self):
+    parser = argparse.ArgumentParser(description='Cortx S3 Setup')
+    parser.add_argument("--cleanup", help='Cleanup S3 accounts and dependencies. Valid arg: both/accounts/dependencies')
```
use "Valid values" instead of "Valid arg"

Also instead of using value as "both", use value "all"

AjinkyaDhumal commented on scripts/provisioning/s3_setup:
```diff
@@ -539,6 +548,15 @@ then
 fi
 
+# Clean up backgrounddelete account
+if [[ $cleanup_both == true ]]
```
Remove the below - 'if [[ $cleanup_dependencies == true ]]' section if its already handled here. Anyways it does not get called now I believe.

amitwac2608 commented on s3psetup/s3psetup/cortx_s3_py_setup.py:
```diff
@@ -0,0 +1,154 @@
+  def dependencies_cleanup(self):
+    """Clean up configs"""
+    log.debug("removing s3 dependencies")
+    cmd = "mv -f /opt/seagate/cortx/auth/resources/authserver.properties /opt/seagate/cortx/auth/resources/authserver.properties.bak"
```
before moving the authserver.properties file, do we need to make sure authserver is down ? @AjinkyaDhumal ?

pratyush-seagate replied:
Please let me know if this existing step needs any change.
FYI:
With new code, this statement should not be run to avoid more issues related to idempotent s3setup (dependencies_cleanup needs to be avoided during idempotent cleanup).

amitwac2608 commented on s3psetup/s3psetup/cortx_s3_py_setup.py:
```diff
@@ -0,0 +1,154 @@
+    """Clean up s3 accounts"""
+    if ldappasswd == None:
+      S3Cipher = CortxS3Cipher(None, True, 12, "openldap")
+      ldappasswd = S3Cipher.get_key()
```
now ldap password will not be cipher key, but we have to read encrypted password from confstore, generate cipher key, decrypt password read from confstore using this key and then use this as ldap password.

amitwac2608 replied:
something like below:

cipherkey=$(s3cipher --generate_key --const_key  openldap 2>/dev/null)
encryptedpassd=$(s3confstore --getkey "cluster>openldap>rootdnpassword" --path "json:///opt/seagate/cortx/s3/conf/s3_confstore.json")

rootdnpasswd=$(s3cipher --decrypt --data $encryptedpassd --key $cipherkey 2>/dev/null)

pratyush-seagate replied:
ok. Is this change already submitted? How can I test it? Because when I ported this code from shell s3_setup, the old approach worked.

pratyush-seagate replied:
Is this change already there in the mini-provisioner head? Because I just rebased and tested latest http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/393/ and seems like the old way of cleanup works fine. What needs to be done here?

pratyush-seagate commented:
Addressing the review caused significant code changes, hence used new custom build to test the changes again. Custom build link: http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/387/

This change needs to be rebased and merged with latest changes from @amitwac2608  from mini-provisioner branch. Will run the final jenkins job after that merge.

pratyush-seagate commented:
Rebase with mini-provisioner head are done. Custom build http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/393/ tested and verified to be working with the two minor adjustments:
1. New cluster block:
{
  "cluster": {
    "node_count" : 1,
    "cluster_id": "dummy-id",
    "cluster_hosts": "localhost"
  }
}

2. New confurl param in s3:config:
/opt/seagate/cortx/s3/bin/s3_setup --updateclusterid --createauthjkspassword --configldap --confurl "json:///opt/seagate/cortx/s3/conf/s3_confstore.json"

sachitanands commented on scripts/provisioning/s3setup/test_s3_setup.sh:
```diff
@@ -0,0 +1,57 @@
+	exit -1
+}
+
+/root/clean_open_ldap_by_s3.sh
```
path can be relative path

pratyush-seagate replied:
Yes, but this is a temporary script and used as a reference on how to test these changes. Anyways I'll address it.

pratyush-seagate replied:
This review is addressed.

sachitanands commented on scripts/provisioning/s3setup/test_s3_setup.sh:
```diff
@@ -0,0 +1,57 @@
+END
+fi
+
+yum install -y --nogpgcheck cortx-motr cortx-s3server openldap-servers openldap-clients &> /dev/null || die_with_error "pkg install failed!"
```
do we need to install haproxy here ?

pratyush-seagate replied:
No not needed for this specific test.

pratyush-seagate commented:
Latest jenkins running at: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/402/

amitwac2608 commented on scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py:
```diff
@@ -0,0 +1,154 @@
+    """Clean up s3 accounts"""
+    if ldappasswd == None:
+      S3Cipher = CortxS3Cipher(None, True, 12, "openldap")
+      ldappasswd = S3Cipher.get_key()
```
Can we make the ldappassword as mandatory option, and pass this from s3_setup script ?

As the get_key() function return cipher key, and that will no longer be the ldappassword.

pratyush-seagate replied:
Ok, this will require some changes in the scripts, and need further testing. Can this be addressed in a followup ticket

pratyush-seagate replied:
I've addressed this change after picking your latest changes from mini-provisioner head. Please review.

amitwac2608 replied:
this section of code is not needed now, remove it in next PR

amitwac2608 commented on scripts/provisioning/s3setup/test_s3_setup.sh:
```diff
@@ -0,0 +1,57 @@
+#!/bin/sh
```
I guess we should make this a python file, as we are trying to get away with shell scripts as much as possible.

pratyush-seagate replied:
Considering this is a temporary script, can we let it be a shell script for the time being? Converting it to python is additional work and I feel we may not need to spend time on this conversion as the base script is already written, working and tested. And this script is also temporary.

amitwac2608 commented on scripts/provisioning/s3setup/test_s3_setup.sh:
```diff
@@ -0,0 +1,57 @@
+./clean_open_ldap_by_s3.sh
+
+if [ ! -e "/opt/seagate/cortx/s3/conf/s3_confstore.json" ]; then
+	tee -a /opt/seagate/cortx/s3/conf/s3_confstore.json << END
```
we should generate this file, not at install path, may be in tmp and delete it later.

pratyush-seagate replied:
Addressed this comment.

pratyush-seagate commented:
@amitwac2608 please review the latest changes which addresses most of the codacy reported issues and all your review comments except convert the temporary shell scripts to python.
The latest custom build http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/418/ is tested and working fine in unit tests including cleanup op. This custom build is prepared after merge to mini-provisioner head. conf file used;
{
 "cluster":
  {
   "cluster_id": "353f5ad1-afa8-4a3f-b55f-cbae4a687107",
   "node_count": "1",
   "server": [
    {
     "hostname": "ssc-vm-1735.colo.seagate.com",
     "roles": "primary,openldap_master",
     "network":
      {
       "data":
        {
         "private_ip": "10.230.240.236"
        }
      }
    }],
   "openldap":
    {
     "sgiampassword": "gAAAAABgBVNaQyfH4EZTl2qFzgg7aFI_vcD_di8FNPQ3-wWIeGteaF8DM08OqOJ2-rQf1bzU8wyapZE1HaoijcE7Ab1Ex_c-Nw==",
     "rootdnpassword": "gAAAAABgBVNaoSC1dWbDewXbt-K39IhOouI6cH79XMmwR2WjnEV1VZclaLiXO2y_f7c2BdMu7XkMHUS3iuTsAay_GUBlMcFiOg=="
    }
  }
}

pratyush-seagate commented:
rebased on mini-provisioner head and http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/417/ started on this branch.

pratyush-seagate commented:
Final jenkins passed at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/419/

amitwac2608 reviewed (approved):
looks good

pratyush-seagate commented:
rebased to latest head of mini-provisioner to include haproxy changes.
https://jts.seagate.com/browse/EOS-16710 created to track the codacy issues reported and not resolved yet in this PR.
